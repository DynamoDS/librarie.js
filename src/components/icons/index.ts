export {default as Arrow} from './Arrow';
