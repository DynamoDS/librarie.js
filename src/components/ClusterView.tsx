/*
      ClusterViewContainer
    +-----------------------------------------------------+
    |  ClusterLeftPane     ClusterRightPane               |
    | +-----------------+ +-----------------------------+ |
    | |    ClusterIcon  | | LibraryItemContainerXxx     | |
    | |       +-------+ | | +-------------------------+ | |
    | |       |       | | | | ...                     | | |
    | |       |       | | | +-------------------------+ | |
    | |       +-------+ | | +-------------------------+ | |
    | |                 | | | ...                     | | |
    | |                 | | +-------------------------+ | |
    | |                 | | +-------------------------+ | |
    | |                 | | | ...                     | | |
    | |                 | | +-------------------------+ | |
    | +-----------------+ +-----------------------------+ |
    +-----------------------------------------------------+
*/

import * as React from "react";
import { LibraryContainer } from "./LibraryContainer";
import { LibraryItem } from "./LibraryItem";
import { ItemData } from "../LibraryUtilities";

export interface ClusterViewProps {
    libraryContainer: LibraryContainer,
    icon: any,
    borderColor: string,
    showExpandableToolTip: boolean,
    childItems: ItemData[]
}

export class ClusterView extends React.Component<ClusterViewProps, undefined> {

    render() {

        let localStyle = {
            borderColor: this.props.borderColor
        };

        return (
            <div className={"ClusterViewContainer"}>
                <div className={"ClusterLeftPane"} style={localStyle}>
                    <img className={"ClusterIcon"} src={this.props.icon} />
                </div>
                <div className={"ClusterRightPane"}>
                    {this.getNestedElements()}
                </div>
            </div>
        );
    }

    getNestedElements() {

        let index = 0;
        return this.props.childItems.map((item: ItemData) => {
            return (<LibraryItem
                key={index++}
                libraryContainer={this.props.libraryContainer}
                showExpandableToolTip={this.props.showExpandableToolTip}
                data={item}
            />);
        });
    }

}
