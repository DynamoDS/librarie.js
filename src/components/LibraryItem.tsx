/*
      LibraryItemContainerXxx(Category/Group/None)
    +---------------------------------------------------+
    |   LibraryItemHeader                               |
    | +-----------------------------------------------+ |
    | | +-----------------+ +-----------------------+ | |
    | | | LibraryItemIcon | | LibraryItemText       | | |
    | | +-----------------+ +-----------------------+ | |
    | +-----------------------------------------------+ |
    |   LibraryItemBody                                 |
    | +-----------------------------------------------+ |
    | |                                               | |
    | |                                               | |
    | +-----------------------------------------------+ |
    +---------------------------------------------------+

*/

import * as React from "react";
import * as ReactDOM from "react-dom";
import { ClusterView } from "./ClusterView";
import { LibraryContainer } from "./LibraryContainer";
import * as LibraryUtilities from "../LibraryUtilities";

export interface LibraryItemProps {
    libraryContainer: any,
    data: LibraryUtilities.ItemData,
    showItemSummary: boolean,
    isLastItem?: boolean,
    onItemWillExpand?: Function
}

export interface LibraryItemState {
    itemSummaryExpanded: boolean,
    expanded: boolean
}

class GroupedItems {

    creates: LibraryUtilities.ItemData[] = [];
    actions: LibraryUtilities.ItemData[] = [];
    queries: LibraryUtilities.ItemData[] = [];
    others: LibraryUtilities.ItemData[] = [];

    constructor(items: LibraryUtilities.ItemData[]) {

        for (let i = 0; i < items.length; i++) {

            switch (items[i].itemType) {
                case "create": this.creates.push(items[i]); break;
                case "action": this.actions.push(items[i]); break;
                case "query": this.queries.push(items[i]); break;
                default: this.others.push(items[i]); break;
            }

            var finalData = {
                "inputParameters": [
                    {
                        "name": "c1",
                        "type": "Color"
                    },
                    {
                        "name": "c2",
                        "type": "Color"
                    }
                ],
                "outputParameters": [
                    "Color"
                ],
                "description": "Construct a Color by combining two input Colors."
            };
        }

        this.creates = LibraryUtilities.sortItemsByText(this.creates);
        this.actions = LibraryUtilities.sortItemsByText(this.actions);
        this.queries = LibraryUtilities.sortItemsByText(this.queries);
        this.others = LibraryUtilities.sortItemsByText(this.others);
    }

    getCreateItems(): LibraryUtilities.ItemData[] { return this.creates; }
    getActionItems(): LibraryUtilities.ItemData[] { return this.actions; }
    getQueryItems(): LibraryUtilities.ItemData[] { return this.queries; }
    getOtherItems(): LibraryUtilities.ItemData[] { return this.others; }
}

export class LibraryItem extends React.Component<LibraryItemProps, LibraryItemState> {

    constructor(props: LibraryItemProps) {
        super(props);

        // All items are collapsed by default, except for section items
        this.state = {
            expanded: this.props.data.expanded,
            itemSummaryExpanded: false
        };

        this.onLibraryItemClicked = this.onLibraryItemClicked.bind(this);
        this.onLibraryItemMouseEnter = this.onLibraryItemMouseEnter.bind(this);
        this.onLibraryItemMouseLeave = this.onLibraryItemMouseLeave.bind(this);
        this.onSectionIconClicked = this.onSectionIconClicked.bind(this);
        this.onSingleChildItemWillExpand = this.onSingleChildItemWillExpand.bind(this);
    }

    // By default all items in search view will be expanded. In search view, 
    // user is still able to expand/unexpand the item, which will toggle the 
    // expansion state. This will make sure that the expansion state of an item
    // in search view will not be affected by the previous user click.
    componentWillReceiveProps(nextProps: LibraryItemProps) {
        if (nextProps.data.expanded !== this.state.expanded) {
            this.setState({ expanded: nextProps.data.expanded });
        }
        //keep the core group defined in layoutspec always expanded.
        //Commented as part of the task : https://jira.autodesk.com/browse/QNTM-2975
        // if(nextProps.data.itemType == "coregroup") {
        //     this.setState({expanded:true});
        // }
    }

    //Afer rendering each Library item, scroll to the expanded item
    componentDidMount() {
        if (this.props.data.expanded && this.props.data.itemType !== "coregroup") {
            //scroll to only that element clicked from search. Determining that element from 
            //other elements is little tricky. The idea here is, the element which has
            //its child elements expanded to false is the actual element clicked from search. Scroll
            //to that element.
            let isThereChildItemsToExpand = this.props.data.childItems.filter((item: any) => {
                return item.expanded == true;
            });
            if (isThereChildItemsToExpand.length == 0) {
                setTimeout(() => {
                    let elem = ReactDOM.findDOMNode(this);
                    elem.scrollIntoView(false);
                }, 0);
            }

        }
    }

    render() {
        if (!this.props.data.visible) {
            return null;
        }

        let nestedElements: JSX.Element = null;
        let clusteredElements: JSX.Element = null;
        let bodyIndentation: JSX.Element = null;
        let header = this.getHeaderElement();

        // visible only nested elements when expanded.
        if (this.state.expanded && this.props.data.childItems.length > 0) {

            // Break item list down into sub-lists based on the type of each item.
            let groupedItems = new GroupedItems(LibraryUtilities.sortItemsByText(this.props.data.childItems));

            // There are some leaf nodes (e.g. methods).
            clusteredElements = this.getClusteredElements(groupedItems);

            // There are intermediate child items.
            nestedElements = this.getNestedElements(groupedItems);
        }

        // Indent one level for clustered and nested elements
        if (this.props.data.itemType !== "section" && (nestedElements || clusteredElements)) {
            bodyIndentation = <div className={"BodyIndentation"} />;
        }

        return (
            <div className={this.getLibraryItemContainerStyle()}>
                {header}
                <div className={"LibraryItemBodyContainer"}>
                    {bodyIndentation}
                    <div className={"LibraryItemBodyElements"}>
                        {clusteredElements}
                        {nestedElements}
                    </div>
                </div>
            </div>
        );
    }

    onImageLoadFail(event: any) {
        event.target.orgSrc = event.target.src;
        event.target.src = require("../resources/icons/default-icon.svg");
    }

    getIconElement(): JSX.Element {
        // If the element type is a section, group, coregroup, class or none an icon shouldn't be displayed
        if (this.props.data.itemType !== "group" && 
        this.props.data.itemType !== "section" && 
        this.props.data.itemType !== "coregroup" &&
        this.props.data.itemType !== "classType" &&
        this.props.data.itemType !== "none") {
            return <img
                className={"LibraryItemIcon"}
                src={this.props.data.iconUrl}
                onError={this.onImageLoadFail}
            />;
        }

         // If it is a section, display the icon (if given) and provide a click interaction.
         //Add-ons have a different style. The icons opacity changes when hovering
        if (this.props.data.itemType === "section" && this.props.data.text == "Add-ons" &&
            this.props.data.iconUrl) {
            return <img
                className={"LibraryAddOnSectionIcon"}
                src={this.props.data.iconUrl}
                onError={this.onImageLoadFail}
                onClick={this.onSectionIconClicked}
            />;
        }

        // If it is a section, display the icon (if given) and provide a click interaction.
        if (this.props.data.itemType === "section" && this.props.data.iconUrl) {
            return <img
                className={"LibraryItemIcon"}
                src={this.props.data.iconUrl}
                onError={this.onImageLoadFail}
                onClick={this.onSectionIconClicked}
            />;
        }

        return null;
    }

    // Show arrow for non-section, non-category and non-leaf items
    getArrowElement(): JSX.Element {
        //no arrow for groups defined in layout spec
        if (this.props.data.itemType === "section" || this.props.data.itemType === "coregroup" ) { 
            return null;
        }

        if (this.props.data.childItems.length == 0) {
            return null;
        };

        let arrowIcon: any = null;

        if (!this.state.expanded) {
            if (this.props.isLastItem) {
                arrowIcon = require("../resources/ui/indent-arrow-right-last.svg");
            } else {
                arrowIcon = require("../resources/ui/indent-arrow-right.svg");
            }
        } else {
            arrowIcon = require("../resources/ui/indent-arrow-down.svg");
        }

        if (this.props.data.itemType == "category") {
            if (!this.state.expanded) {
                arrowIcon = require("../resources/ui/indent-arrow-right-wo-lines.svg");
            }
            else {
                arrowIcon = require("../resources/ui/indent-arrow-down-wo-lines.svg");
            }

            return <img className={"CategoryArrow"} src={arrowIcon} onError={this.onImageLoadFail} />;

        }

        return <img className={"Arrow"} src={arrowIcon} onError={this.onImageLoadFail} />;
    }

    getHeaderElement(): JSX.Element {
        let arrow = this.getArrowElement();
        let iconElement = this.getIconElement();
        let parameters: JSX.Element = null;

        if (this.props.data.parameters && this.props.data.parameters.length > 0 && this.props.data.childItems.length == 0) {
            parameters = <div className="LibraryItemParameters">{this.props.data.parameters}</div>;
        }

        if (this.props.data.showHeader) {
            return (
                <div className={this.getLibraryItemHeaderStyle()} onClick={this.onLibraryItemClicked}
                    onMouseEnter={this.onLibraryItemMouseEnter} onMouseLeave={this.onLibraryItemMouseLeave}>
                    {arrow}
                    {this.props.data.itemType === "section" ? null : iconElement}
                    <div className={this.getLibraryItemTextStyle()}>{this.props.data.text}</div>
                    {parameters}
                    {this.props.data.itemType === "section" ? iconElement : null}
                </div>
            );
        }

        return null;
    }

    getLibraryItemContainerStyle(): string {
        switch (this.props.data.itemType) {
            case "section":
                return "LibraryItemContainerSection";
            case "category":
                return "LibraryItemContainerCategory";
            case "group":
            case "coregroup":
                return "LibraryItemContainerGroup";
            default:
                return "LibraryItemContainerNone";
        }
    }

    getLibraryItemHeaderStyle(): string {
        switch (this.props.data.itemType) {
            case "section":
                return "LibrarySectionHeader";
            default:
                return "LibraryItemHeader";
        }
    }

    getLibraryItemTextStyle(): string {
        switch (this.props.data.itemType) {
            case "group":
            case "coregroup":
            case "section":
                return "LibraryItemGroupText";
            default:
                return "LibraryItemText";
        }
    }

    getNestedElements(groupedItems: GroupedItems): JSX.Element {

        let regularItems = groupedItems.getOtherItems();
        if (regularItems.length <= 0) {
            return null; // No item to be generated.
        }

        let index = 0;
        return (
            <div className={"LibraryItemBody"}>
                {
                    // 'getNestedElements' method is meant to render all other 
                    // types of items except ones of type create/action/query.
                    regularItems.map((item, i) => {
                        let isLastItem = false;

                        if (i == regularItems.length - 1) {
                            isLastItem = true;
                        }
                        return (<LibraryItem
                            key={index++}
                            libraryContainer={this.props.libraryContainer}
                            data={item}
                            showItemSummary={this.props.showItemSummary}
                            isLastItem={isLastItem}
                            onItemWillExpand={(args: any) => {
                                this.onSingleChildItemWillExpand();
                                this.props.libraryContainer.scrollToExpandedItem(args)
                            }}
                        />);
                    })
                }
            </div>
        );
    }

    getClusteredElements(groupedItems: GroupedItems): JSX.Element {

        let createMethods = groupedItems.getCreateItems();
        let actionMethods = groupedItems.getActionItems();
        let queryMethods = groupedItems.getQueryItems();

        let createCluster = null;
        if (createMethods.length > 0 && createMethods.some(item => item.visible)) {
            createCluster = (<ClusterView
                libraryContainer={this.props.libraryContainer}
                icon={require("../resources/icons/library-create.svg")}
                borderColor="#62895b" /* green */
                showItemSummary={this.props.showItemSummary}
                childItems={createMethods} />);
        }

        let actionCluster = null;
        if (actionMethods.length > 0 && actionMethods.some(item => item.visible)) {
            actionCluster = (<ClusterView
                libraryContainer={this.props.libraryContainer}
                icon={require("../resources/icons/library-action.svg")}
                borderColor="#ad5446" /* red */
                showItemSummary={this.props.showItemSummary}
                childItems={actionMethods} />);
        }

        let queryCluster = null;
        if (queryMethods.length > 0 && queryMethods.some(item => item.visible)) {
            queryCluster = (<ClusterView
                libraryContainer={this.props.libraryContainer}
                icon={require("../resources/icons/library-query.svg")}
                borderColor="#4b9dbf" /* blue */
                showItemSummary={this.props.showItemSummary}
                childItems={queryMethods} />);
        }

        if ((!createCluster) && (!actionCluster) && (!queryCluster)) {
            return null; // No cluster should be generated.
        }

        return (
            <div className={"LibraryItemBody"}>
                {createCluster}
                {actionCluster}
                {queryCluster}
            </div>
        );
    }

    onLibraryItemClicked() {
        //https://jira.autodesk.com/browse/QNTM-2975
        //Add-ons section is always expanded.
        if(this.props.data.text == "Add-ons") return;
        // Toggle expansion state.
        let currentlyExpanded = this.state.expanded;

        if (this.props.data.childItems.length > 0 && !currentlyExpanded && this.props.onItemWillExpand) {
            this.props.onItemWillExpand(ReactDOM.findDOMNode(this));
        }

        this.setState({ expanded: !currentlyExpanded });

        //auto expand the coregroup elements
        //commenting as part of the task : https://jira.autodesk.com/browse/QNTM-2975
        // if(this.props.data.itemType === "category" ) {
        //     this.props.data.childItems.forEach((item: any) => {
        //         if(item.itemType == "coregroup"){
        //             item.expanded = true;
        //         }
        //     });
        // }

        let libraryContainer = this.props.libraryContainer;
        if (this.props.data.childItems.length == 0) {
            libraryContainer.raiseEvent(libraryContainer.props.libraryController.ItemClickedEventName,
                this.props.data.contextData);
        }
    }

    onSectionIconClicked(event: any) {
        let libraryContainer = this.props.libraryContainer;
        libraryContainer.raiseEvent(libraryContainer.props.libraryController.SectionIconClickedEventName, this.props.data.text);
        event.stopPropagation(); // Prevent the onClick event of its parent item from being called.
    }

    // Collapse all child items when one of the child items is expanded
    onSingleChildItemWillExpand() {
        for (let item of this.props.data.childItems) {
            item.expanded = false;
        }
        this.setState({ expanded: true }); // Make the current item (parent) expanded.
    }

    onLibraryItemMouseLeave() {
        let libraryContainer = this.props.libraryContainer;
        if (this.props.data.childItems.length == 0) {
            let mouseLeaveEvent = libraryContainer.props.libraryController.ItemMouseLeaveEventName;
            libraryContainer.raiseEvent(mouseLeaveEvent, { data: this.props.data.contextData });
        }
    }

    onLibraryItemMouseEnter() {
        let libraryContainer = this.props.libraryContainer;
        if (this.props.data.childItems.length == 0) {
            let rec = ReactDOM.findDOMNode(this).getBoundingClientRect();
            let mouseEnterEvent = libraryContainer.props.libraryController.ItemMouseEnterEventName;
            libraryContainer.raiseEvent(mouseEnterEvent, { data: this.props.data.contextData, rect: rec, element:  ReactDOM.findDOMNode(this) });
        }
    }
}
