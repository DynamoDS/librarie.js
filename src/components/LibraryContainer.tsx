/// <reference path="../../node_modules/@types/node/index.d.ts" />

require("../resources/LibraryStyles.css");
require("../resources/fonts/font-awesome-4.7.0/css/font-awesome.min.css");

import * as React from "react";
import { LibraryController } from "../entry-point";
import { LibraryItem } from "./LibraryItem";
import { SearchView } from "./SearchView";
import * as LibraryUtilities from "../LibraryUtilities";

declare var boundContainer: any; // Object set from C# side.

export interface LibraryContainerProps {
    libraryController: LibraryController,
    defaultSectionString: string,
    miscSectionString: string
}

export interface LibraryContainerStates {
    inSearchMode: boolean
}

export class LibraryContainer extends React.Component<LibraryContainerProps, LibraryContainerStates> {

    loadedTypesJson: any = null;
    layoutSpecsJson: any = null;

    generatedSections: LibraryUtilities.ItemData[] = null;
    searchCategories: string[] = [];

    constructor(props: LibraryContainerProps) {
        super(props);

        // Bind function prototypes to the object instance.
        this.setLoadedTypesJson = this.setLoadedTypesJson.bind(this);
        this.setLayoutSpecsJson = this.setLayoutSpecsJson.bind(this);
        this.refreshLibraryView = this.refreshLibraryView.bind(this);
        this.onSearchModeChanged = this.onSearchModeChanged.bind(this);

        // Set handlers after methods are bound.
        this.props.libraryController.setLoadedTypesJsonHandler = this.setLoadedTypesJson;
        this.props.libraryController.setLayoutSpecsJsonHandler = this.setLayoutSpecsJson;
        this.props.libraryController.refreshLibraryViewHandler = this.refreshLibraryView;

        this.state = { inSearchMode: false };
    }

    setLoadedTypesJson(loadedTypesJson: any, append: boolean = true): void {

        if (!loadedTypesJson) {
            throw new Error("Parameter 'loadedTypesJson' must be supplied");
        }

        if (!loadedTypesJson.loadedTypes || (!Array.isArray(loadedTypesJson.loadedTypes))) {
            throw new Error("'loadedTypesJson.loadedTypes' must be a valid array");
        }

        // If there is no existing loadedTypesJson object, or the call
        // is meant to replace the existing one, then assign it and bail.
        if (!this.loadedTypesJson || (!append)) {
            this.loadedTypesJson = loadedTypesJson;
            return;
        }

        // To append to the existing 'loadedTypesJson.loadedTypes object' (merge both arrays).
        Array.prototype.push.apply(this.loadedTypesJson.loadedTypes, loadedTypesJson.loadedTypes);
    }

    setLayoutSpecsJson(layoutSpecsJson: any, append: boolean = true): void {

        if (!layoutSpecsJson) {
            throw new Error("Parameter 'layoutSpecsJson' must be supplied");
        }

        if (!layoutSpecsJson.sections || (!Array.isArray(layoutSpecsJson.sections))) {
            throw new Error("'layoutSpecsJson.sections' must be a valid array");
        }

        // If there is no existing layoutSpecsJson object, or the call
        // is meant to replace the existing one, then assign it and bail.
        if (!this.layoutSpecsJson || (!append)) {
            this.layoutSpecsJson = layoutSpecsJson;
            return;
        }

        // Otherwise, recursively replace/append each section.
        LibraryUtilities.updateSections(this.layoutSpecsJson, layoutSpecsJson);
    }

    refreshLibraryView(): void {

        this.generatedSections = LibraryUtilities.buildLibrarySectionsFromLayoutSpecs(
            this.loadedTypesJson, this.layoutSpecsJson,
            this.props.defaultSectionString, this.props.miscSectionString);

        // Obtain the categories from each section to be added into the filtering options for search
        for (let section of this.generatedSections) {
            for (let childItem of section.childItems)
                this.searchCategories.push(childItem.text);
        }

        // Just to force a refresh of UI.
        this.setState({ inSearchMode: this.state.inSearchMode });
    }

    raiseEvent(name: string, params?: any | any[]) {
        this.props.libraryController.raiseEvent(name, params);
    }

    onSearchModeChanged(inSearchMode: boolean) {
        this.setState({ inSearchMode: inSearchMode });
    }

    render() {
        if (!this.generatedSections) {
            return (<div>This is LibraryContainer</div>);
        }

        try {
            let sections: JSX.Element[] = null;
            const searchView = <SearchView onSearchModeChanged={this.onSearchModeChanged}
                libraryContainer={this} sections={this.generatedSections} categories={this.searchCategories} />;

            if (!this.state.inSearchMode) {
                let index = 0;
                sections = this.generatedSections.map(data => <LibraryItem key={index++} libraryContainer={this} data={data} />)
            }

            return (
                <div className="LibraryContainer">
                    {searchView}
                    <div className="LibraryItemContainer">
                        {sections}
                    </div>
                </div>
            );
        } catch (exception) {
            return (<div>Exception thrown: {exception.message}</div>);
        }
    }
}
