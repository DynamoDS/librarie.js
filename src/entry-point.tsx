import * as React from "react";
import * as ReactDOM from "react-dom";

import { LibraryContainer } from "./components/LibraryContainer";
import { JsonDownloader } from "./LibraryUtilities";
import { Reactor, Event } from "./EventHandler";

export function CreateJsonDownloader(jsonUrls: string[], callback: Function) {
    return new JsonDownloader(jsonUrls, callback);
}

export function CreateLibraryController() {
    return new LibraryController();
}

export class LibraryController {

    ItemClickedEventName = "itemClicked";
    ItemMouseEnterEventName = "itemMouseEnter";
    ItemMouseLeaveEventName = "itemMouseLeave";

    reactor: Reactor = null;

    constructor() {
        this.on = this.on.bind(this);
        this.raiseEvent = this.raiseEvent.bind(this);
        this.createLibraryByElementId = this.createLibraryByElementId.bind(this);
        this.createLibraryContainer = this.createLibraryContainer.bind(this);

        this.reactor = new Reactor();
    }

    on(eventName: string, callback: Function) {
        this.reactor.registerEvent(eventName, callback);
    }

    raiseEvent(name: string, params?: any | any[]) {
        this.reactor.raiseEvent(name, params);
    }

    createLibraryByElementId(htmlElementId: string, layoutSpecsJson: any, loadedTypesJson: any) {
        let htmlElement = document.getElementById(htmlElementId);
        return ReactDOM.render(<LibraryContainer
            libraryController={this}
            loadedTypesJson={loadedTypesJson}
            layoutSpecsJson={layoutSpecsJson} />, htmlElement);
    }

    createLibraryContainer(layoutSpecsJson: any, loadedTypesJson: any) {
        return (<LibraryContainer
            libraryController={this}
            layoutSpecsJson={layoutSpecsJson}
            loadedTypesJson={loadedTypesJson} />);
    }
}
