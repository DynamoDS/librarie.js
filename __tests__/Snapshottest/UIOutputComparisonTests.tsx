/**
 * @jest-environment jsdom
 */

import * as React from 'react';
import * as LibraryEntryPoint from '../../src/entry-point';
import { LibraryContainer } from '../../src/components/LibraryContainer';
import { LibraryItem } from '../../src/components/LibraryItem';
import { ItemData } from "../../src/LibraryUtilities";
import { shallow, configure } from 'enzyme';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as Adapter from 'enzyme-adapter-react-16';
import * as chai from 'chai';

configure({adapter: new Adapter()});

describe("LibraryContainer", function () {

  it("Test UI rendering of single component of Library Item", function () {
    // Test data to create LibraryItem  
    // Parent item that will hold the child items
    let data = new ItemData("");
    data.text = "TestItem";
    data.itemType = "none";

    // Create child items and link with parent
    for (var i = 0; i < 2; i++) {
      let item = new ItemData("");
      item.text = "Child" + i;
      item.itemType = "category";
      data.appendChild(item);
    }

    // Create "LibraryContainer" to pass as an argument for creation of "LibraryItem"
    let libContainer = LibraryEntryPoint.CreateLibraryController();
    let libraryItem = shallow(<LibraryItem libraryContainer={libContainer} data={data} showItemSummary={false} />);
    // When running for the first time creates a snapshot in __snapshots__ folder
    // To compare with the existing snapshot for subsequent running. 
    expect(toJson(libraryItem)).toMatchSnapshot();
  });

  it("Test UI rendering of Library Item and child components", function () {
    // Test data to create LibraryItem  
    // Parent item that will hold the child items
    let data = new ItemData("");
    data.text = "TestItem";
    data.itemType = "none";

    // create child items and link with parent
    for (var i = 0; i < 2; i++) {
      let item = new ItemData("");
      item.text = "Child" + i;
      item.itemType = "category";
      data.appendChild(item);
    }
    // Create "LibraryContainer" to pass as an argument for creation of "LibraryItem"
    let libContainer = LibraryEntryPoint.CreateLibraryController();
    // Render with mount to test child components
    let libraryItem = mount(<LibraryItem libraryContainer={libContainer} data={data} showItemSummary={false} />);
    expect(toJson(libraryItem)).toMatchSnapshot();  // toJson serializes the output    
  });
  it("Demonstrate testing UIitems loads correctly from static data", function () {
    // This is demo test to show how to test the LibrayUI is rendered correctly
    // Dynamo dynamcally generates the data in the libraryUI
    // So this test does not ensure all the required items are loaded there
    let layoutSpecsJson = require("../../docs/LayoutSpecs.json");
    let loadedTypesJson = require("../../docs/RawTypeData.json");
    let libController = LibraryEntryPoint.CreateLibraryController();
    // Render with mount to test child components
    const tree =mount(libController.createLibraryContainer()
      ); // required to dump the html version of librarie.js
      // create the library Controleer with the static data in the libraryui project 
    libController.setLoadedTypesJson(loadedTypesJson, false);
    libController.setLayoutSpecsJson(layoutSpecsJson, false);
    libController.refreshLibraryView();
    let libContainer = LibraryEntryPoint.CreateLibraryController();
    // Search for all LibraryItems      
    let text = tree.find('div.LibraryItemText');
    expect(toJson(text)).toMatchSnapshot();
  });
});

