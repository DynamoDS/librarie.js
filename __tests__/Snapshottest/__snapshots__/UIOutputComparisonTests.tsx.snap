// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LibraryContainer Demonstrate testing UIitems loads correctly from static data 1`] = `
Array [
  <div
    className="LibraryItemText"
>
    Display
</div>,
  <div
    className="LibraryItemText"
>
    Geometry
</div>,
  <div
    className="LibraryItemText"
>
    ImportExport
</div>,
  <div
    className="LibraryItemText"
>
    Input
</div>,
  <div
    className="LibraryItemText"
>
    List
</div>,
  <div
    className="LibraryItemText"
>
    Math
</div>,
  <div
    className="LibraryItemText"
>
    Script
</div>,
  <div
    className="LibraryItemText"
>
    String
</div>,
  <div
    className="LibraryItemText"
>
    Clockwork
</div>,
  <div
    className="LibraryItemText"
>
    DynamoText
</div>,
]
`;

exports[`LibraryContainer Test UI rendering of Library Item and child components 1`] = `
<LibraryItem
  data={
    ItemData {
      "childItems": Array [
        ItemData {
          "childItems": Array [],
          "contextData": "",
          "description": "",
          "expanded": false,
          "iconUrl": "",
          "itemType": "category",
          "keywords": Array [
            "",
          ],
          "parameters": "",
          "pathToItem": Array [],
          "showHeader": true,
          "text": "Child0",
          "visible": true,
          "weight": 0,
        },
        ItemData {
          "childItems": Array [],
          "contextData": "",
          "description": "",
          "expanded": false,
          "iconUrl": "",
          "itemType": "category",
          "keywords": Array [
            "",
          ],
          "parameters": "",
          "pathToItem": Array [],
          "showHeader": true,
          "text": "Child1",
          "visible": true,
          "weight": 0,
        },
      ],
      "contextData": "",
      "description": "",
      "expanded": false,
      "iconUrl": "",
      "itemType": "none",
      "keywords": Array [
        "",
      ],
      "parameters": "",
      "pathToItem": Array [],
      "showHeader": true,
      "text": "TestItem",
      "visible": true,
      "weight": 0,
    }
  }
  libraryContainer={
    LibraryController {
      "DefaultSectionName": "default",
      "FilterCategoryEventName": "filterCategoryChange",
      "ItemClickedEventName": "itemClicked",
      "ItemMouseEnterEventName": "itemMouseEnter",
      "ItemMouseLeaveEventName": "itemMouseLeave",
      "ItemSummaryExpandedEventName": "itemSummaryExpanded",
      "MiscSectionName": "Miscellaneous",
      "RefreshLibraryViewRequestName": "refreshLibraryView",
      "SearchTextUpdatedEventName": "searchTextUpdated",
      "SectionIconClickedEventName": "sectionIconClicked",
      "createLibraryByElementId": [Function],
      "createLibraryContainer": [Function],
      "on": [Function],
      "raiseEvent": [Function],
      "reactor": Reactor {
        "events": Array [],
      },
      "refreshLibraryView": [Function],
      "refreshLibraryViewHandler": null,
      "registerRequestHandler": [Function],
      "request": [Function],
      "requestHandler": Object {},
      "searchLibraryItemsHandler": null,
      "setLayoutSpecsJson": [Function],
      "setLayoutSpecsJsonHandler": null,
      "setLoadedTypesJson": [Function],
      "setLoadedTypesJsonHandler": null,
    }
  }
>
  <div
    className="LibraryItemContainerNone"
  >
    <div
      className="LibraryItemHeader"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <img
        className="Arrow"
        onError={[Function]}
        src="test-file-stub"
      />
      <div
        className="LibraryItemText"
      >
        TestItem
      </div>
    </div>
    <div
      className="LibraryItemBodyContainer"
    >
      <div
        className="LibraryItemBodyElements"
      />
    </div>
  </div>
</LibraryItem>
`;

exports[`LibraryContainer Test UI rendering of single component of Library Item 1`] = `
<div
  className="LibraryItemContainerNone"
>
  <div
    className="LibraryItemHeader"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <img
      className="Arrow"
      onError={[Function]}
      src="test-file-stub"
    />
    <div
      className="LibraryItemText"
    >
      TestItem
    </div>
  </div>
  <div
    className="LibraryItemBodyContainer"
  >
    <div
      className="LibraryItemBodyElements"
    />
  </div>
</div>
`;
